File: Assets/Bundles/Server/Resources/TS/Services/Global/Block/DenyRegionService.lua
-- Compiled with unity-ts v2.1.0-75
local Reflect = require("Shared/rbxts_include/node_modules/@easy-games/flamework-core/out/init").Reflect
local Service = require("Shared/rbxts_include/node_modules/@easy-games/flamework-core/out/init").Service
local ServerSignals = require("Server/TS/ServerSignals").ServerSignals
local Network = require("Shared/TS/Network").Network
local Task = require("Shared/TS/Util/Task").Task
-- * Deny region snapshot send delay.
local SNAPSHOT_DELAY = 0.5
local DenyRegionService
do
	DenyRegionService = setmetatable({}, {
		__tostring = function()
			return "DenyRegionService"
		end,
	})
	DenyRegionService.__index = DenyRegionService
	function DenyRegionService.new(...)
		local self = setmetatable({}, DenyRegionService)
		return self:constructor(...) or self
	end
	function DenyRegionService:constructor()
		self.denyRegionIdCounter = 0
		self.trackedDenyRegions = {}
		self.denyVoxelPositions = {}
	end
	function DenyRegionService:OnStart()
		-- Cancel block placed if voxel position is inside of a deny region.
		ServerSignals.BeforeBlockPlaced:ConnectWithPriority(0, function(event)
			if self:InDenyRegion(event.pos) then
				event:setCancelled(true)
			end
		end)
		-- Send deny region snapshot to late joiners.
		ServerSignals.PlayerJoin:connect(function(event)
			Task:Delay(SNAPSHOT_DELAY, function()
				Network.ServerToClient.DenyRegionSnapshot.Server:FireClient(event.player.clientId, self.trackedDenyRegions)
			end)
		end)
	end
	function DenyRegionService:CreateDenyRegion(origin, size)
		local _object = {
			origin = origin,
			size = size,
		}
		local _left = "id"
		local _original = self.denyRegionIdCounter
		self.denyRegionIdCounter += 1
		_object[_left] = tostring(_original)
		local newTrackedDenyRegion = _object
		table.insert(self.trackedDenyRegions, newTrackedDenyRegion)
		do
			local x = origin.x - math.floor(size.x / 2)
			local _shouldIncrement = false
			while true do
				if _shouldIncrement then
					x += 1
				else
					_shouldIncrement = true
				end
				if not (x <= origin.x + math.ceil(size.x / 2)) then
					break
				end
				do
					local y = origin.y
					local _shouldIncrement_1 = false
					while true do
						if _shouldIncrement_1 then
							y += 1
						else
							_shouldIncrement_1 = true
						end
						if not (y <= origin.y + size.y) then
							break
						end
						do
							local z = origin.z - math.floor(size.z / 2)
							local _shouldIncrement_2 = false
							while true do
								if _shouldIncrement_2 then
									z += 1
								else
									_shouldIncrement_2 = true
								end
								if not (z <= origin.z + math.ceil(size.z / 2)) then
									break
								end
								local denyPosition = Vector3.new(x, y, z)
								self.denyVoxelPositions[denyPosition] = true
							end
						end
					end
				end
			end
		end
		Network.ServerToClient.DenyRegionCreated.Server:FireAllClients(newTrackedDenyRegion)
	end
	function DenyRegionService:InDenyRegion(position)
		local _denyVoxelPositions = self.denyVoxelPositions
		local _position = position
		return _denyVoxelPositions[_position] ~= nil
	end
end
-- (Flamework) DenyRegionService metadata
Reflect.defineMetadata(DenyRegionService, "identifier", "Bundles/Server/Services/Global/Block/DenyRegionService@DenyRegionService")
Reflect.defineMetadata(DenyRegionService, "flamework:implements", { "$:flamework@OnStart" })
Reflect.decorate(DenyRegionService, "$:flamework@Service", Service, { {} })
return {
	DenyRegionService = DenyRegionService,
}
-- ----------------------------------
-- ----------------------------------

File: Assets/Bundles/Imports/Core/Core~/src/Server/Services/Block/DenyRegionService.ts
import { OnStart, Service } from "@easy-games/flamework-core";
import { CoreServerSignals } from "Server/CoreServerSignals";
import { CoreNetwork } from "Shared/CoreNetwork";
import { DenyRegionDto } from "Shared/DenyRegion/DenyRegionMeta";
import { SignalPriority } from "Shared/Util/Signal";
import { Task } from "Shared/Util/Task";

/** Deny region snapshot send delay. */
const SNAPSHOT_DELAY = 0.5;

@Service({})
export class DenyRegionService implements OnStart {
	/** Sequential deny region id counter. */
	private denyRegionIdCounter = 0;
	/** Tracked deny regions for recreation on client. */
	private trackedDenyRegions: DenyRegionDto[] = [];
	/** Deny voxel positions. */
	private denyVoxelPositions = new Set<Vector3>();

	OnStart(): void {
		/* Cancel block placed if voxel position is inside of a deny region. */
		CoreServerSignals.BeforeBlockPlaced.ConnectWithPriority(SignalPriority.HIGHEST, (event) => {
			if (this.InDenyRegion(event.pos)) event.SetCancelled(true);
		});
		/* Send deny region snapshot to late joiners. */
		CoreServerSignals.PlayerJoin.Connect((event) => {
			Task.Delay(SNAPSHOT_DELAY, () => {
				CoreNetwork.ServerToClient.DenyRegionSnapshot.Server.FireClient(
					event.player.clientId,
					this.trackedDenyRegions,
				);
			});
		});
	}

	/**
	 * Creates a deny region at `origin` of size `size`.
	 * @param origin The deny region origin.
	 * @param size The deny region size.
	 */
	public CreateDenyRegion(origin: Vector3, size: Vector3): void {
		const newTrackedDenyRegion: DenyRegionDto = {
			origin: origin,
			size: size,
			id: tostring(this.denyRegionIdCounter++),
		};
		this.trackedDenyRegions.push(newTrackedDenyRegion);
		for (let x = origin.x - math.floor(size.x / 2); x <= origin.x + math.ceil(size.x / 2); x++) {
			for (let y = origin.y; y <= origin.y + size.y; y++) {
				for (let z = origin.z - math.floor(size.z / 2); z <= origin.z + math.ceil(size.z / 2); z++) {
					const denyPosition = new Vector3(x, y, z);
					this.denyVoxelPositions.add(denyPosition);
				}
			}
		}
		CoreNetwork.ServerToClient.DenyRegionCreated.Server.FireAllClients(newTrackedDenyRegion);
	}

	/**
	 * Checks whether or not `position` is inside of a deny region.
	 * @param position A voxel position.
	 * @returns Whether or not `position` is inside of a deny region.
	 */
	public InDenyRegion(position: Vector3): boolean {
		return this.denyVoxelPositions.has(position);
	}
}
